cot/src/response/into_response.rs

```diff
@@ -0,0 +1,258 @@
+use std::borrow::Cow;
+use std::convert::Infallible;
+
+use bytes::{Bytes, BytesMut};
+use cot::headers::HTML_CONTENT_TYPE;
+use cot::response::{RESPONSE_BUILD_FAILURE, Response};
+use cot::{Body, StatusCode};
+use http;
+
+use crate::html::Html;
+
+/// Trait for generating responses.
+/// Types that implement `IntoResponse` can be returned from handlers.
+///
+/// # Implementing `IntoResponse`
+///
+/// You generally shouldn't have to implement `IntoResponse` manually, as cot
+/// provides implementations for many common types.
+///
+/// However, it might be necessary if you have a custom error type that you want
+/// to return from handlers.
+
+pub trait IntoResponse {
+    /// Create a response.
+    #[must_use]
+    fn into_response(self) -> Response;
+
+    fn with_header<K, V>(self, key: K, value: V) -> Response
+    where
+        K: Into<http::HeaderName>,
+        V: Into<http::HeaderValue>,
+        Self: Sized,
+    {
+        let key = key.into();
+        let value = value.into();
+
+        let mut response = self.into_response();
+        response.headers_mut().append(key, value);
+        response
+    }
+
+    fn with_content_type<V>(self, content_type: V) -> Response
+    where
+        V: Into<http::HeaderValue>,
+        Self: Sized,
+    {
+        let mut response = self.into_response();
+        response
+            .headers_mut()
+            .insert(http::header::CONTENT_TYPE, content_type.into());
+        response
+    }
+
+    fn with_status(self, status: StatusCode) -> Response
+    where
+        Self: Sized,
+    {
+        let mut response = self.into_response();
+        *response.status_mut() = status;
+        response
+    }
+
+    fn with_body(self, body: impl Into<Body>) -> Response
+    where
+        Self: Sized,
+    {
+        let mut response = self.into_response();
+        *response.body_mut() = body.into();
+        response
+    }
+}
+macro_rules! impl_into_response_for_type_and_mime {
+    ($ty:ty, $mime:expr) => {
+        impl IntoResponse for $ty {
+            fn into_response(self) -> Response {
+                let mut res = Body::from(self).into_response();
+                res.headers_mut().insert(
+                    http::header::CONTENT_TYPE,
+                    http::HeaderValue::from_static($mime.as_ref()),
+                );
+                res
+            }
+        }
+    };
+}
+
+// General implementations
+
+impl IntoResponse for () {
+    fn into_response(self) -> Response {
+        Body::empty().into_response()
+    }
+}
+
+impl IntoResponse for Infallible {
+    fn into_response(self) -> Response {
+        match self {}
+    }
+}
+
+impl<R, E> IntoResponse for Result<R, E>
+where
+    R: IntoResponse,
+    E: IntoResponse,
```

Comment:
What do you feel about changing this line to the below instead:

```suggestion
    E: Into<cot::Error>,
```

? I think it would make more sense to allow propagating errors easily, rather than creating a response out of them (because we already have dedicated ways to create responses out of errors built-in in the framework). Right now this seems like it was just copy-pasted from Axum, but our "core concepts" are slightly different, and I don't think this impl fits well in our case.

Reply:
Nope, that does not make sense, unfortunately. Within the `RequestHandler`s we do not know the type of the results of handlers apart from the fact that they implement `IntoResponse`, so that `Result` becomes a generic `Response`, so we would not be able to handle it anyway. Having `E: IntoResponse` ensures that whatever the user decides to return on error, it can be a valid response.

Reply:
Part of the reason I asked this is because I don't really see a good use case for this `impl`. Right now, the `RequestHandler` `impl` requires that the return type is `cot::Result<_>`. In order to use this `IntoResponse` `impl`, you'd have to return `cot::Result<Result<_, _>>`. Why would you want to do this?

And, as I mentioned in another review comment, I think we shouldn't throw away our built-in error handling. By ignoring the fact that the `Error` type is, well, an error, we won't ever display our error pages that are there to benefit the framework users and simplify the development.